libs/typesignatures/src/main/java/io/crate/typesignatures/TypeSignaturesASTVisitor.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.typesignatures;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+import io.crate.typesignatures.antlr.v4.TypeSignaturesBaseVisitor;
+import io.crate.typesignatures.antlr.v4.TypeSignaturesParser;
+
+public class TypeSignaturesASTVisitor extends TypeSignaturesBaseVisitor<Node> {
```

Comment:
Instead of first building up an AST and then converting that to the actual type signatures, would it be possible to immediate build up the type signatures in one pass?

Reply:
It's a good idea, although that would require to move this into the server package.